File: extension6/CoffeeHouseWithSupplyManager.py
from typing import Optional

from extension6.SupplyManagerWorkPlace import SupplyManagerWorkPlace
from root.CoffeeHouse import CoffeeHouse
from root.Ingredient import Ingredient
from root.Product import Product


class CoffeeHouseWithSupplyManager(CoffeeHouse):
    _supplyManagerWorkPlace: Optional[SupplyManagerWorkPlace]

    def __init__(self):
        super().__init__()
        self._supplyManagerWorkPlace = None

    def addIngredient(self, ingredient: Ingredient, amount: int) -> None:
        super().addIngredient(ingredient, amount)
        self._supplyManagerWorkPlace.addIngredientInitialAmount(ingredient, amount)

    def addDessert(self, dessert: Product, amount: int) -> None:
        super().addDessert(dessert, amount)
        self._supplyManagerWorkPlace.addProductInitialAmount(dessert, amount)

    def addSupplyManagerWorkPlace(self, supplyManagerWorkPlace: SupplyManagerWorkPlace) -> None:
        if self._supplyManagerWorkPlace is None:
            self._supplyManagerWorkPlace = supplyManagerWorkPlace

    def takeProduct(self, product: Product) -> None:
        if not product.isNeedPrepare():
            currentAmount = self._dessertMap.get(product, 0)
            if currentAmount > 0:
                self._dessertMap[product] -= 1
                if self._dessertMap[product] < 1:
                    self._supplyManagerWorkPlace.requestProductRenew(product)

    def takeIngredient(self, ingredient: Ingredient, amount: int) -> bool:
        currentAmount = self._ingredientMap.get(ingredient, 0)
        if (currentAmount - amount) >= 0:
            self._ingredientMap[ingredient] = currentAmount - amount
            if self._ingredientMap[ingredient] < 0:
                self._supplyManagerWorkPlace.requestIngredientRenew(ingredient)
            return True
        return False

File: cplusplus/ext6/CoffeeHouseWithSupplyManager.cpp
#include "CoffeeHouseWithSupplyManager.hpp"

CoffeeHouseWithSupplyManager::CoffeeHouseWithSupplyManager() : CoffeeHouse()
{
}

void CoffeeHouseWithSupplyManager::addIngredient(Ingredient *ingredient, int amount)
{
	CoffeeHouse::addIngredient(ingredient, amount);
	this->supplyManager->addIngredientInitialAmount(ingredient, amount);
}

void CoffeeHouseWithSupplyManager::addDessert(Product *dessert, int amount)
{
	CoffeeHouse::addDessert(dessert, amount);
	this->supplyManager->addProductInitialAmount(dessert, amount);
}

void CoffeeHouseWithSupplyManager::addSupplyManager(SupplyManager* supplyManager)
{
	if (this->supplyManager == nullptr)
		this->supplyManager = supplyManager;
}

void CoffeeHouseWithSupplyManager::takeProduct(Product *product)
{
	if (!product->getIsNeedPrepare())
	{
		auto it = dessertMap.find(product);
		int currentAmount;
		if (it != dessertMap.end())
			currentAmount = it->second;
		else
			currentAmount = 0;
		if (currentAmount > 0)
		{
			dessertMap[product]--;
			if (this->dessertMap[product] < 1)
				if (this->supplyManager->getManagerState())
				{
					long time = this->supplyManager->renewProduct(product, this->currentTime);
					this->currentTime += time;
				}
		}
	}
}

bool CoffeeHouseWithSupplyManager::takeIngredient(Ingredient *ingredient, int amount)
{
	auto it = ingredientMap.find(ingredient);
	int currentAmount;
	if (it != ingredientMap.end())
		currentAmount = it->second;
	else
		currentAmount = 0;
	if (currentAmount - amount >= 0)
	{
		ingredientMap[ingredient] -= amount;
		if (this->ingredientMap[ingredient] < 0)
			if (this->supplyManager->getManagerState())
			{
				long time = this->supplyManager->renewIngredient(ingredient, this->currentTime);
				this->currentTime += time;
			}
		return true;
	}
	return false;
}

